packages/react-dom-bindings/src/server/ReactDOMServerFormatConfig.js

```diff
@@ -62,18 +62,68 @@ import isArray from 'shared/isArray';
 // E.g. this can be used to distinguish legacy renderers from this modern one.
 export const isPrimaryRenderer = true;
 
+// Shape of argument objects for each instruction writer.
+//  Note that the formatted chunk arguments do not include
+//  enclosing '"'s
+export type CompletedSegmentInstructionWriterArgs = {
+  write: (Chunk | PrecomputedChunk) => boolean,
+  responseState: ResponseState,
+  containerID: Chunk,
+  placeholderID: Chunk,
+};
+export type CompletedBoundaryInstructionWriterArgs = {
+  write: (Chunk | PrecomputedChunk) => boolean,
+  responseState: ResponseState,
+  boundaryID: null | PrecomputedChunk,
+  contentSegmentID: Chunk,
+};
+export type ClientRenderBoundaryInstructionWriterArgs = {
+  write(Chunk | PrecomputedChunk): boolean,
+  responseState: ResponseState,
+  boundaryID: SuspenseBoundaryID,
+  errorDigest: Chunk | null,
+  errorMessage: Chunk | null,
+  errorComponentStack: Chunk | null,
+};
+export type InstructionWriter = {
+  writeCompletedSegmentInstruction(
+    args: CompletedSegmentInstructionWriterArgs,
+  ): boolean,
+
+  writeCompletedBoundaryInstruction(
+    args: CompletedBoundaryInstructionWriterArgs,
+  ): boolean,
+
+  writeClientRenderBoundaryInstruction(
+    args: ClientRenderBoundaryInstructionWriterArgs,
+  ): boolean,
+};
+
+const defaultInstructionWriter: InstructionWriter = {
+  writeCompletedSegmentInstruction: defaultWriteCompletedSegmentInstruction,
+  writeCompletedBoundaryInstruction: defaultWriteCompletedBoundaryInstruction,
+  writeClientRenderBoundaryInstruction: defaultWriteClientRenderBoundaryInstruction,
+};
+
+// State for default instruction writer, which streams executable scripts
+type DefaultInstructionWriterState = {
+  startInlineScript: PrecomputedChunk,
+  sentCompleteSegmentFunction: boolean,
+  sentCompleteBoundaryFunction: boolean,
+  sentClientRenderFunction: boolean,
+};
+
 // Per response, global state that is not contextual to the rendering subtree.
 export type ResponseState = {
   bootstrapChunks: Array<Chunk | PrecomputedChunk>,
-  startInlineScript: PrecomputedChunk,
-  placeholderPrefix: PrecomputedChunk,
-  segmentPrefix: PrecomputedChunk,
+  placeholderPrefix: string,
+  segmentPrefix: string,
   boundaryPrefix: string,
   idPrefix: string,
   nextSuspenseID: number,
-  sentCompleteSegmentFunction: boolean,
-  sentCompleteBoundaryFunction: boolean,
-  sentClientRenderFunction: boolean, // We allow the legacy renderer to extend this object.
+  instructionWriter: InstructionWriter,
+  instructionWriterState: mixed,
```

Comment:
Not sure what the right flow type to use for this is. 

Ideally, the writer and writerState types should be associated (e.g. InstructionWriter is templated on some writer state). But from my noob Flow understanding, this might complicate passing around ResponseState. 

Would love feedback on making this better (if the rest of the code looks reasonable)